v3/rfcs/dx/20240620-simplify-subgraph-getting-started-and-federation.md

```diff
@@ -0,0 +1,47 @@
+# DX: simplify subgraphs for getting started and federation workflows
+
+Status: Review required
+
+## subgraph init command
+What this is: as a part of making it easier for people to work through the getting started guide, or to scaffold project files, allow the user to go from scratch to an API with only CLI commands.
+
+init-ing a supergraph or connector will create files for both local and cloud environments. This is about what happens for the subgraphs.
+
+```bash
+# creates a subgraph folder and subgraph.yaml file
+ddn subgraph init app
+
+# add subgraph.yaml to the supergraph.yaml file and create an env file for the env
+ddn supergraph add subgraph --supergraph supergraph.local.yaml --subgraph app/subgraph.yaml
+touch app/.env.app.local
```

Comment:
From a primitives PoV, `touch` to create a .env file makes sense. But this is gonna be different for UNIX and powershell. Do during subgraph init, can we by default initialize a `.env.app.local` file (and add it to the `subgraph.yaml` as well)? The rationale is we init a subgrph, do local dev and then go to cloud, so a local .env file is absolutely needed here, so why not create it?

Reply:
`ddn connector-link add` takes a --configure-host flag. Add a --env-file flag here as well. Creates the env file if not present.

Now we can remove the touch command.

Reply:
Why don't we initialize a `subgraph.cloud.yaml` and subgraph.local.yaml` when we initialize a subgraph along with the corresponding .env files? The files will be the same except for the .env file.

The advantages of doing so are:

1. We generate local and cloud files for supergraph and connector. We also have to generate local and cloud file for the globals subgraph.yaml. Doing the same when initializing a subgraph would keep the behaviour consistent with everything else we are doing.
2. It would make the commands much more simpler as users don't **need** to pass a `--env-file` flag. Advanced users can still do it if they want but for simple use cases and getting started, not having to pass an additional flag would keep things simpler and easier to grasp. The fact that it is more consistent as mentioned in (1) would also make it more easier to grasp.

The only drawback is that we'll have 2 files that are the same except for the .env file. However since this is how it's done in all other cases and since this will eliminate the need to add an extra flag in all the commands, this seems like the more simpler approach.

Reply:
> The command should be ddn subgraph add

IMO, this sounds like we are adding something to the subgraph. I feel like `ddn supergraph add subgraph` as in the original makes more sense although it's more verbose. Or were you talking about something else?

Reply:
Let's go with
* single subgraph.yaml
* scaffold both .env.app.local and .env.app.cloud

This does not introduce additional subgraph.yaml files with no differences, but also makes the model clear to the user.

Reply:
ddn subgraph add --supergraph
is along the lines of
ddn model add --subgraph
and so is consistent

The alternative here could be
ddn supergraph subgraph add
Which is more verbose and doesn't really add much value